ostd/src/sync/rcu/non_null.rs

```diff
@@ -0,0 +1,241 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! This module provides a trait and some auxiliary types to help abstract and
+//! work with non-null pointers.
+
+use alloc::sync::Weak;
+use core::{marker::PhantomData, mem::ManuallyDrop, ops::Deref, ptr::NonNull};
+
+use crate::prelude::*;
+
+/// A trait that abstracts pointers that are non-null.
+///
+/// The most typical examples smart pointer types like `Box<T>` and `Arc<T>`,
+/// which can be converted to and from the raw pointer type of `*const T`.
+///
+/// # Safety
+///
+/// This trait must be implemented correctly (according to the doc comments for
+/// each method). Types like [`Rcu`] rely on this assumption to safely use the
+/// raw pointers.
+///
+/// [`Rcu`]: super::Rcu
+pub unsafe trait NonNullPtr: Send + 'static {
+    /// The target type that this pointer refers to.
+    // TODO: allow ?Sized
+    type Target;
+
+    /// A type that behaves just like a shared references to the `NonNullPtr`.
+    type Ref<'a>: Deref<Target = Self>
+    where
+        Self: 'a;
+
+    /// Converts to a raw pointer.
+    ///
+    /// Each call to `into_raw` must be paired with a call to `from_raw`
+    /// in order to avoid memory leakage.
+    fn into_raw(self) -> NonNull<Self::Target>;
+
+    /// Converts back from a raw pointer.
+    ///
+    /// # Safety
+    ///
+    /// 1. The raw pointer must have been previously returned by a call to
+    ///    `into_raw`.
+    /// 2. The raw pointer must not be used after calling `from_raw` if the
+    ///    pointer is required to be unique (e.g., `Box<T>`).
+    unsafe fn from_raw(ptr: NonNull<Self::Target>) -> Self;
+
+    /// Obtains a shared reference to the original pointer.
+    ///
+    /// # Safety
+    ///
+    /// The original pointer must outlive the lifetime parameter `'a`, and during `'a`
+    /// no mutable references to the pointer will exist.
+    unsafe fn raw_as_ref<'a>(raw: NonNull<Self::Target>) -> Self::Ref<'a>;
+
+    /// Converts a shared reference to a pointer.
+    fn ref_as_raw(ptr_ref: Self::Ref<'_>) -> NonNull<Self::Target>;
+}
+
+/// A type that represents `&'a Box<T>`.
+#[derive(PartialEq, Debug)]
```

Comment:
It's not clear what `PartialEq` means on `BoxRef` and `ArcRef`. Does it mean the pointer is the same or the pointee is the same?